### ISSUE TITLE
JSONExtract functions do not support arbitrary precision numeric types

### ISSUE THREAD

**Issue by marcioapm** [ClickHouse/ClickHouse#41056]
JSONExtract functions do not support arbitrary precision numeric types like Decimal or Integer128.
It also doesn't convert strings, the only way to represent such large numbers in JSON, to the intended numeric type.

Some JSON encoders will encode values stored as 64-bit in the source representation as strings, others will do so only if the value cannot be represented precisely as a double.
There is no efficient way to deal with this in Clickhouse at the moment.
The only way it to use JSON_VALUE but this is very slow as it does a lot of redundant work and always converts the result to string.

Maybe we need a new family of functions that support this dynamic nature of JSON and will try harder to perform coercing and match the desired type.

```
SELECT JSONExtract('{"a":1,"b":"99"}', 'Tuple(a Int128, b Decimal128(6))');`

Code: 43. DB::Exception: Received from 127.0.0.1:9000. DB::Exception: Function JSONExtract doesn't support the return type schema: Int128: While processing JSONExtract('{"a":1,"b":"99"}', 'Tuple(a Int128, b Decimal128(6))'). (ILLEGAL_TYPE_OF_ARGUMENT)
```

Would be nice to have for example something like:
```
SELECT CoercingJSONExtract('{"a":1, "b":"99"}', 'Tuple(a Int128, b Int128)');
(1, 99)


SELECT CoercingJSONExtract('{"a":1, "b":"99"}', 'a', 'Int128');
1

SELECT CoercingJSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Int128');
0

SELECT CoercingJSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Nullable(Int128)');
null
 ```

**Comment by Avogar** [comment]
There is no need to create a new function `CoercingJSONExtract ` for it. It would be better to add support for these types for `JSONExtract`. Seems like after `JSONExtract` was added we added some new data types and dind't add support for them in `JSONExtract`

**Comment by Avogar** [comment]
BTW, `Decimal` types are supported, but for some reason they doesn't work always:
```
avogar-dev :) SELECT JSONExtract('{"a":1}', 'a', 'Decimal128(6)')

SELECT JSONExtract('{"a":1}', 'a', 'Decimal128(6)')

Query id: dc7fce95-3e83-41b5-a8f8-b3e7edd5a5ba

┌─JSONExtract('{"a":1}', 'a', 'Decimal128(6)')─┐
│                                            0 │
└──────────────────────────────────────────────┘

1 row in set. Elapsed: 0.012 sec.

avogar-dev :) SELECT JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')

SELECT JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')

Query id: f23b4874-29bd-40e3-a8f5-60a4d07fb5f8

┌─JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')─┐
│                                            1.1 │
└────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.012 sec.
```
It would be great to fix it.

**Comment by marcioapm** [comment]
Exactly, it's because there is no coercing.

> BTW, `Decimal` types are supported, but for some reason they doesn't work always:
> 
> ```
> avogar-dev :) SELECT JSONExtract('{"a":1}', 'a', 'Decimal128(6)')
> 
> SELECT JSONExtract('{"a":1}', 'a', 'Decimal128(6)')
> 
> Query id: dc7fce95-3e83-41b5-a8f8-b3e7edd5a5ba
> 
> ┌─JSONExtract('{"a":1}', 'a', 'Decimal128(6)')─┐
> │                                            0 │
> └──────────────────────────────────────────────┘
> 
> 1 row in set. Elapsed: 0.012 sec.
> 
> avogar-dev :) SELECT JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')
> 
> SELECT JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')
> 
> Query id: f23b4874-29bd-40e3-a8f5-60a4d07fb5f8
> 
> ┌─JSONExtract('{"a":1.1}', 'a', 'Decimal128(6)')─┐
> │                                            1.1 │
> └────────────────────────────────────────────────┘
> 
> 1 row in set. Elapsed: 0.012 sec.
> ```
> 
> It would be great to fix it.

Yes, the reason they don't work it's because there is no coercing.

I also would prefer to just adjust the current functions and make them universally useful, but it will be a behaviour change.
Shall I prepare a PR for it?

**Comment by Avogar** [comment]
I would prefer to adjust the current functions, but let me also ask my collegues

**Comment by Avogar** [comment]
Let's extend generic `JSONExtract` function, which has the data type argument.
It will be great if you can make a PR with it!

**Comment by marcioapm** [comment]
If we accept the change in behaviour, then to make it work as one would expect, more intuitively, then I think it would be best to fix them all. I see no reason to leave the specifically typed functions "broken".

**Comment by marcioapm** [comment]
Sounds good. Will have a PR ready soonish.

**Comment by marcioapm** [comment]
[PR submitted #41502](https://github.com/ClickHouse/ClickHouse/pull/41502)

I would appreciate some guidance getting the PR merged.
There is a bunch of work left to make `JSONExtract` more usable in practice - I will attempt to fix those too over time.

**Comment by UnamedRus** [comment]
Merged in 22.11

```sql
SELECT JSONExtract('{"a":1,"b":"99"}', 'Tuple(a Int128, b Decimal128(6))')


┌─JSONExtract('{"a":1,"b":"99"}', 'Tuple(a Int128, b Decimal128(6))')─┐
│ (1,99)                                                              │
└─────────────────────────────────────────────────────────────────────┘

SELECT JSONExtract('{"a":1,"b":"99"}', 'a', 'Int128')


┌─JSONExtract('{"a":1,"b":"99"}', 'a', 'Int128')─┐
│                                              1 │
└────────────────────────────────────────────────┘

SELECT JSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Int128')


┌─JSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Int128')─┐
│                                                                 0 │
└───────────────────────────────────────────────────────────────────┘



SELECT JSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Nullable(Int128)')

┌─JSONExtract('{"a":1, "b":"99","c":"some_string"}', 'c', 'Nullable(Int128)')─┐
│                                                                        ᴺᵁᴸᴸ │
└─────────────────────────────────────────────────────────────────────────────┘

```
